File: Neetcode/Stack/largestrecthist.py
# https://neetcode.io/problems/largest-rectangle-in-histogram
# https://leetcode.com/problems/largest-rectangle-in-histogram/

from typing import List


class Solution:
    def largestRectangleArea(self, heights: List[int]) -> int:
        prev = ret = heights[0]
        st = [(prev, 0)]
        for i, curr in enumerate(heights[1:]):
            if curr > prev:
                st.append((curr, i + 1))
            elif curr < prev:
                while st and st[-1][0] > curr:
                    th, ti = st.pop()
                    ret = max(ret, th * (i - ti + 1))
                st.append((curr, ti))
            prev = curr

        l = len(heights)
        for h, i in st:
            ret = max(ret, h * (l - i))
        return ret

File: Neetcode/Stack/largestrecthist.java
package Neetcode.Stack;

import java.util.Stack;

public class largestrecthist {

}

class Solution {
  public int largestRectangleArea(int[] heights) {
    if (heights == null || heights.length == 0) {
      return 0;
    }

    Stack<Integer> st = new Stack<>();
    st.push(-1);
    int max = 0;

    for (int i = 0; i < heights.length; i++) {
      while (st.size() != 1 && heights[i] < heights[st.peek()]) {
        max = Math.max(max, heights[st.pop()] * ((i - 1) - st.peek()));
      }
      st.push(i);
    }

    while (st.size() != 1) {
      max = Math.max(max, heights[st.pop()] * ((heights.length - 1) - st.peek()));
    }

    return max;
  }
}